File: services/client/src/protocol/bets_records.go
package protocol

import (
	"bytes"
	"encoding/binary"
	"fmt"

	lottery "github.com/7574-sistemas-distribuidos/tp-nivelador/src/model"
)

// / mantains shared logic for BetsPayload and WinnersPayload
type betRecordList struct {
	Records   []lottery.Bet
	batchSize int
}

const (
	documentLen  = 4
	numberLen    = 2
	fnLenLen     = 1
	lnLenLen     = 1
	birthdateLen = 8
	headerLen    = documentLen + numberLen + fnLenLen + lnLenLen + birthdateLen // 16
)

// Adds a bet to the list. Returns an error if the batch is full.
func (b *betRecordList) AddBet(bet lottery.Bet) error {
	if len(b.Records) >= b.batchSize {
		return &FullBatchError{}
	}
	b.Records = append(b.Records, bet)
	return nil
}

func (b *betRecordList) MarshalPayload() ([]byte, error) {
	// Pre-calculate total size to avoid per-bet allocations
	total := 2 // 2-byte count header
	for _, bet := range b.Records {
		total += headerLen + len(bet.FirstName) + len(bet.LastName)
	}

	buf := make([]byte, total)
	binary.BigEndian.PutUint16(buf[0:2], uint16(len(b.Records)))
	offset := 2

	for _, bet := range b.Records {
		offset = marshalBetRecordInto(buf, offset, bet)
	}
	return buf, nil
}

func (p *betRecordList) UnmarshalPayload(data []byte) error {
	if len(data) < 2 {
		return &InvalidPayloadError{mesage: "payload too short"}
	}
	count := binary.BigEndian.Uint16(data[0:2])
	offset := 2

	for i := 0; i < int(count); i++ {
		bet, new_offset := UnmarshalBetRecord(offset, data)
		p.Records = append(p.Records, bet)
		offset = new_offset
	}
	if len(p.Records) != int(count) {
		return &InvalidPayloadError{mesage: fmt.Sprintf("bets count mismatch: expected %d, got %d", count, len(p.Records))}
	}
	return nil
}

type BetsPayload struct {
	betRecordList
}

func CreateBetsPayload(batchSize int) *BetsPayload {
	return &BetsPayload{betRecordList{batchSize: batchSize}}
}

func (p *BetsPayload) Type() MessageType { return BETS }

type WinnersPayload struct {
	betRecordList
}

func (p *WinnersPayload) Type() MessageType { return WINNERS }

func CreateWinnersPayload(batchSize int) *WinnersPayload {
	return &WinnersPayload{betRecordList{batchSize: batchSize}}
}
func (p *WinnersPayload) GetWinners() []lottery.Bet {
	return p.Records
}

// marshalBetRecordInto writes a bet record into buf starting at offset and returns the new offset.
func marshalBetRecordInto(buf []byte, offset int, bet lottery.Bet) int {
	lenName := len(bet.FirstName)
	lenLastName := len(bet.LastName)

	binary.BigEndian.PutUint32(buf[offset:], uint32(bet.Document))
	offset += documentLen

	binary.BigEndian.PutUint16(buf[offset:], uint16(bet.Number))
	offset += numberLen

	buf[offset] = byte(lenName)
	offset += fnLenLen

	buf[offset] = byte(lenLastName)
	offset += lnLenLen

	copy(buf[offset:offset+birthdateLen], bet.Birthdate) // padea con 0x00 si falta
	offset += birthdateLen

	copy(buf[offset:], bet.FirstName)
	offset += lenName

	copy(buf[offset:], bet.LastName)
	offset += lenLastName

	return offset
}
func UnmarshalBetRecord(offset int, payload []byte) (lottery.Bet, int) {
	document := int(binary.BigEndian.Uint32(payload[offset : offset+4]))
	offset += 4

	number := int(binary.BigEndian.Uint16(payload[offset : offset+2]))
	offset += 2

	nameLength := int(payload[offset])
	offset++

	lastNameLength := int(payload[offset])
	offset++

	birthdate := string(bytes.TrimRight(payload[offset:offset+8], "\x00"))
	offset += 8

	name := string(payload[offset : offset+nameLength])
	offset += nameLength

	lastName := string(payload[offset : offset+lastNameLength])
	offset += lastNameLength

	bet := lottery.Bet{
		//AgencyId:  agencyID,
		FirstName: name,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthdate,
		Number:    number,
	}
	return bet, offset
}

File: services/server/src/server/protocol/bets_records.py
from __future__ import annotations
from .messages import Message, BETS, WINNERS
from lottery.bet import Bet

LEN_BET_MIN = 16


class BetsMessage(Message):
    def __init__(self, agency_id: int, batch_size: int = 1024):
        self.type = BETS
        super().__init__(agency_id, batch_size)
        self.bets = []

    def add_bet(self, bet):
        if len(self.bets) >= self.batch_size:
            raise ValueError("cannot add bet: batch is full")
        self.bets.append(bet)

    def _marshall_payload(self) -> bytes:
        payload = len(self.bets).to_bytes(2, byteorder='big')
        for bet in self.bets:
            payload += bet.document.to_bytes(4, byteorder='big')
            payload += bet.number.to_bytes(2, byteorder='big')
            name_bytes = bet.first_name.encode('utf-8')
            last_name_bytes = bet.last_name.encode('utf-8')
            payload += len(name_bytes).to_bytes(1, byteorder='big')
            payload += len(last_name_bytes).to_bytes(1, byteorder='big')
            payload += bet.birthdate.encode('utf-8').ljust(8, b'\x00')
            payload += name_bytes
            payload += last_name_bytes
        return payload

    @classmethod
    def _unmarshall_payload(cls, agency_id: int, payload: bytes) -> BetsMessage:
        msg = cls(agency_id)
        bets_count = int.from_bytes(payload[0:2], byteorder='big')
        offset = 2
        for _ in range(bets_count):
            document = int.from_bytes(payload[offset:offset + 4], byteorder='big'); offset += 4
            number = int.from_bytes(payload[offset:offset + 2], byteorder='big'); offset += 2
            name_length = payload[offset]; offset += 1
            last_name_length = payload[offset]; offset += 1
            birthdate = payload[offset:offset + 8].decode('utf-8').rstrip('\x00'); offset += 8
            name = payload[offset:offset + name_length].decode('utf-8'); offset += name_length
            last_name = payload[offset:offset + last_name_length].decode('utf-8'); offset += last_name_length

            bet = Bet(agency_id, name, last_name, document, birthdate, number)
            msg.bets.append(bet)  # ojo: append directo, NO add_bet (ver nota abajo)
        return msg

class WinnersMessage(BetsMessage):
    type = WINNERS